src/ros_robot_localization_listener.cpp

```diff
@@ -0,0 +1,278 @@
+    // - - - - - - - - - - - - - - - - - -
+
+    // First get the transform from base to target
+    geometry_msgs::TransformStamped base_to_target_transform;
+    try
+    {
+      base_to_target_transform = tf_buffer_.lookupTransform(base_frame_id_,
+                                                            frame_id,
+                                                            ros::Time(time),
+                                                            ros::Duration(0.1)); // TODO: magic number
+    }
+    catch ( tf2::LookupException e )
+    {
+      ROS_WARN_STREAM("Ros Robot Localization Listener: Could not look up transform: " << e.what());
+      return false;
+    }
+
+    // And convert it to an eigen Affine transformation
+    Eigen::Affine3d target_pose_base;
+    tf::transformMsgToEigen(base_to_target_transform.transform, target_pose_base);
+
+    // Then convert the base pose to an Eigen Affine transformation
+    Eigen::Vector3d base_position(estimator_state.state(StateMemberX),
+                                  estimator_state.state(StateMemberY),
+                                  estimator_state.state(StateMemberZ));
+
+    Eigen::AngleAxisd roll_angle(estimator_state.state(StateMemberRoll), Eigen::Vector3d::UnitX());
+    Eigen::AngleAxisd pitch_angle(estimator_state.state(StateMemberPitch), Eigen::Vector3d::UnitY());
+    Eigen::AngleAxisd yaw_angle(estimator_state.state(StateMemberYaw), Eigen::Vector3d::UnitZ());
+
+    Eigen::Quaterniond base_orientation(yaw_angle * pitch_angle * roll_angle);
+
+    Eigen::Affine3d base_pose(Eigen::Translation3d(base_position) * base_orientation);
+
+    // Now we can calculate the transform from odom to the requested frame (target)...
+    Eigen::Affine3d target_pose_odom = base_pose * target_pose_base;
+
+    // ... and put it in the output state
+    state(StateMemberX) = target_pose_odom.translation().x();
+    state(StateMemberY) = target_pose_odom.translation().y();
+    state(StateMemberZ) = target_pose_odom.translation().z();
+
+    Eigen::Vector3d ypr = target_pose_odom.rotation().eulerAngles(2, 1, 0);
+
+    state(StateMemberRoll)  = ypr[2];
+    state(StateMemberPitch) = ypr[1];
+    state(StateMemberYaw)   = ypr[0];
+
+    // Now let's calculate the twist of the target frame
+    Twist base_vel, target_vel_base;
+    base_vel.linear = Eigen::Vector3d(estimator_state.state(StateMemberVx),
+                                      estimator_state.state(StateMemberVy),
+                                      estimator_state.state(StateMemberVz));
+    base_vel.angular = Eigen::Vector3d(estimator_state.state(StateMemberVroll),
+                                       estimator_state.state(StateMemberVpitch),
+                                       estimator_state.state(StateMemberVyaw));
+
+    Eigen::Vector3d target_position_base = target_pose_base.translation();
+    target_vel_base.linear = base_vel.linear + base_vel.angular.cross(target_position_base);
```

Comment:
No rotation is necessary, because all vectors here are expressed in the base-fixed frame.